drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1073 @@
+        }
+
+        return drmModeAtomicAddProperty(req, plane_id, prop_id, value);
+}
+
+/*
+ * modeset_atomic_modeset() is also new. It is used to atomically set some
+ * properties to our connector, CRTC and plane objects. This function is used to
+ * setup the modeset before we start drawing.
+ *
+ * Notice that we can't perform an atomic commit without a configured
+ * framebuffer. The intention here is not to draw (the framebuffer is
+ * zeroed), we just want to setup everything before we start to draw.
+ *
+ * After we have every property set in our own objects, we call
+ * drmModeAtomicCommit(). This function internally creates a struct
+ * drm_mode_atomic object that receives all the properties we want to change.
+ * In the end, a single IOCTL will be called using this drm_mode_atomic object:
+ * DRM_IOCTL(fd, DRM_IOCTL_MODE_ATOMIC, &atomic)
+ */
+
+static int modeset_atomic_modeset(int fd, struct modeset_dev *dev)
+{
+	drmModeAtomicReq *req;
+	uint32_t blob_id;
+	int ret, flags;
+	struct drm_object *plane = &dev->plane;
+	struct modeset_buf *buf = &dev->bufs[dev->front_buf ^ 1];
+
+	req = drmModeAtomicAlloc();
+
+	/* prepare our connector object to set the CRTC id that
+	 * the connector is using */
+	if (set_connector_property(req, &dev->connector, dev->conn_id, 
+							   "CRTC_ID", dev->crtc_id) < 0)
+		return -1;
+
+	/* Creates a blob property that receives dev->mode value. Then, the CRTC
+	 * object property "MODE_ID" points to the id of the blob. This usually
+	 * happens for properties that are not simple types. In this case, dev->mode
+	 * is a struct. But we could have a property that expects the id of a blob
+	 * that holds an array, for instance.
+	 */
+	if (drmModeCreatePropertyBlob(fd, &dev->mode, sizeof(dev->mode), &blob_id) != 0)
+		return -1;
+	if (set_crtc_property(req, &dev->crtc, dev->crtc_id, "MODE_ID", blob_id) < 0)
+		return -1;
+
+	/* prepare our CRTC object to be set as active */
+	if (set_crtc_property(req, &dev->crtc, dev->crtc_id, "ACTIVE", 1) < 0)
+		return -1;
+
+	/* prepare to set properties of the plane related to the CRTC and the
+	 * framebuffer */
+	set_plane_property(req, plane, dev->plane_id, "FB_ID", buf->fb);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_ID", dev->crtc_id);
+	set_plane_property(req, plane, dev->plane_id, "SRC_X", 0);
+	set_plane_property(req, plane, dev->plane_id, "SRC_Y", 0);
+	set_plane_property(req, plane, dev->plane_id, "SRC_W", buf->width << 16);
+	set_plane_property(req, plane, dev->plane_id, "SRC_H", buf->height << 16);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_X", 0);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_Y", 0);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_W", buf->width);
+	set_plane_property(req, plane, dev->plane_id, "CRTC_H", buf->height);
+
+	/* Set the connector, CRTC and plane properties in a single IOCTL. Here we
+	 * leave it as a blocking call, since we can't start to do page-flips before
+	 * setting everything up.
```

Comment:
Actually, you can ask for a pageflip event and call this in a non-blocking fashion for the non-test commit.

With the legacy KMS API, there was no way to ask for a non-blocking modeset. With atomic we gain the ability to do non-blocking modesets in addition to the old non-blocking flips: `DRM_MODE_PAGE_FLIP_EVENT | DRM_MODE_ATOMIC_NONBLOCK`

With atomic API, if we set  `ALLOW_MODESET`, the kernel driver is free to do a modeset or a pageflip, whatever works.

If we do not set  `ALLOW_MODESET`, the kernel driver is forbidden to cause a glitch, meaning that quite likely it is forbidden from doing a modeset. If the state we submit requires a modeset, the atomic commit will fail in that case.

Reply:
Cool, removed this comment. I thought it was a good idea not to change this to use a non-blocking commit, since it would make the code more complex. But if you disagree, I can change that.